cloudfoundry/cf.go

```diff
@@ -0,0 +1,214 @@
+	authClient  httpclient.AuthenticatedClient
+	apiUrl      string
+	accessToken string
+}
+
+func GetOauthToken(client httpclient.HttpClient, apiUrl string, username string, password string) (string, error) {
+	request, err := http.NewRequest("GET", apiUrl+"/v2/info", nil)
+	if err != nil {
+		return "", fmt.Errorf("unable to build http request: %s", err)
+	}
+	request.Header.Add("Accept", "application/json")
+	var infoResp infoResponse
+
+	err = do(client, request, &infoResp)
+	if err != nil {
+		return "", fmt.Errorf("/v2/info failure: %s", err)
+	}
+
+	request, err = http.NewRequest("GET", infoResp.AuthorisationEndpoint+"/login", nil)
+	if err != nil {
+		return "", err
+	}
+	request.Header.Add("Accept", "application/json")
+	var loginResp loginResponse
+	err = do(client, request, &loginResp)
+	if err != nil {
+		return "", fmt.Errorf("/login failure: %s", err)
+	}
+
+	body := fmt.Sprintf("grant_type=password&password=%s&scope=&username=%s", password, username)
+	request, err = http.NewRequest("POST", loginResp.Links.Login+"/oauth/token", strings.NewReader(body))
+	if err != nil {
+		return "", err
+	}
+	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	request.Header.Add("Accept", "application/json")
+	accessToken := "Basic " + base64.StdEncoding.EncodeToString([]byte("cf:"))
+	request.Header.Add("Authorization", accessToken)
+	var tokenResp tokenResponse
+	err = do(client, request, &tokenResp)
+	if err != nil {
+		return "", fmt.Errorf("/outh/token failure: %s", err)
+	}
+
+	return tokenResp.AccessToken, nil
+}
+
+func NewClient(authClient httpclient.AuthenticatedClient, apiUrl, accessToken string) Client {
+	return &client{
+		authClient:  authClient,
+		apiUrl:      apiUrl,
+		accessToken: accessToken,
+	}
+}
+
+func (cf *client) GetServices(serviceName string) (services []Service, err error) {
+	var servicesResponse listServicesResponse
+	err = cf.get(fmt.Sprintf("/v2/services?q=label:%s", serviceName), &servicesResponse)
+	services = servicesResponse.Resources
+	return
+}
+
+func (cf *client) GetServicePlans(serviceGuid string, resultsPerPage int) (servicePlans chan ServicePlan, errorChannel chan error) {
+	servicePlans = make(chan ServicePlan, resultsPerPage)
+	errorChannel = make(chan error, 1)
+
+	endpoint := fmt.Sprintf("/v2/services/%s/service_plans?results-per-page=%d", serviceGuid, resultsPerPage)
+
+	go func() {
+		defer close(servicePlans)
+		defer close(errorChannel)
+
+		for endpoint != "" {
+			var servicePlanResponse listServicePlansResponse
+			err := cf.get(endpoint, &servicePlanResponse)
+			if err != nil {
+				errorChannel <- err
+				return
+			}
+
+			for _, servicePlan := range servicePlanResponse.Resources {
+				servicePlans <- servicePlan
+			}
+
+			endpoint = servicePlanResponse.NextUrl
+		}
+	}()
+
+	return
+}
+
+func (cf *client) GetServicePlanInstances(servicePlanGuid string, resultsPerPage int) (servicePlanInstances chan ServiceInstance, errorChannel chan error) {
```

Comment:
Why not simply return a slice of service plan instances? If `resultsPerPage` is set to a reasonably high value, then there will typically be just one remote request and so any performance gain from concurrency will be minimal. Returning a channel just seems to complicate the signature, and therefore the caller, and make testing more difficult.

Reply:
Yep that's very true - I'll swap it around.

Reply:
Thanks!

Reply:
I misread this the first time around as 'service plans' rather than 'service plan instances'. I just pushed a change to `func (cf *client) GetServicePlans(serviceGuid string) (servicePlans []ServicePlan, err error)`, since I think it's totally reasonable to assume the number of plans is small and we can just grab them all in one go.

For service plan instances, we either have to make the same assumption that we can suck all instances into memory at once, or reaper has to know how to deal with paging and make multiple requests through the cf client, passing a page number. I went down the channel route to keep that kind of thing out of reaper, since I preferred reaper just being able to work on a queue of stuff, rather than having to worry about managing the fetches as well. But yes, it means the function signature and testing is a bit less straightforward than if we just returned `([]ServiceInstance, error)`. What do you reckon?

Reply:
I don't feel strongly. Unless you can be reasonably confident that the number of service plan instances for a given service plan is going to be fairly small (less than 100, for instance), then you may as well stick with the current code. You may be interested in [this article](https://ewencp.org/blog/golang-iterators/index.html) which contrasts various options including channels, closures, and iterators. If one of these approaches which avoid spawning a goroutine look appealing, it might be worth an hour to see how it pans out, but I'll leave that to your judgement as the current code looks fairly safe and I don't imagine service instance reaper having a long history of functional enhancements.

Reply:
Now I look at it again, reaper has control of the number of results per page. I think I should have hidden that away in `cf.go`.

Reply:
That's a nice article - good to see those approaches compared side by side.

> What I’d really like to see is some sort of range adapter: provide an interface in the builtin package that lets you use any iterable in a range expression. That should be the unifying interface and allow you to use any of these interchangeably and get really clean calling code. It could just be the usual HasNext()/Next() interface seen in other languages. I assume the issue with this is that there is no way to specify those generically.

^^ this is how I would have done it in Java, without really thinking about it. In Go, the primitives make this pipelining approach so easy, whereas to do the same in Java you'd end up with `BlockingQueue` and a thread pool - feels so overkill you'd never bother for something like this.  Funny to think how the language has affected my decisions with interesting side-effects.